### PR TITLE
Autosave

### PR DESCRIPTION
**Pull request by Macroz**
Implements autosave from #2767 

Autosave happens "on change", when a second has passed from last activity.

# Notes

- [x] When user session has expired, autosave does not actually autosave after the login redirect. 
  - OK, session is not likely to expire with keep alives and while you are editing the autosave is triggered after 1s so there shouldn't be anything significant lost
- [x] Should the event description be changed "Alice saved a draft." -> "Alice updated the application."
  - Let's do it after we clean up the regular save draft and this is the default behavior (#3037).
- [x] Move attachment save feedback to actions box without the scrolling. We could make it use the autosave or at least imitate the behavior (view does not jump away).
- [x] Attachment remove does not trigger autosave.
- [x] Clean up the flash message code. There are many similar functions. (separate issue #3038)
- [x] Browser test both autosave and regular save

# Screenshots

autosave1
autosave2
autosave3

# Checklist for author

Remove items that aren't applicable, check items that are done.

## Reviewability
- [x] Link to issue

## Backwards compatibility
- [x] Events are backwards compatible _or_ have migrations (events are the same)
- [x] Config is backwards compatible (autosave is disabled by default except in dev)

## Documentation
- [x] Update changelog if necessary
- [ ] Update docs/ (if applicable)
- [ ] Update manual/ (if applicable)
- [x] ADR for major architectural decisions or experiments (existing ADR 017)
- [x] New config options in config-defaults.edn

## Testing
- [x] Valuable features are integration / browser / acceptance tested automatically

## Follow-up
- [ ] New tasks are created for pending or remaining tasks

### CODE REVIEW

**Comment by aatkin** [comment]
DUO field is not autosaved currently (although it is still under works, the feature could be implemented at least in initial capacity)

**Comment by aatkin** [comment]
> > DUO field is not autosaved currently (although it is still under works, the feature could be implemented at least in initial capacity)
> 
> Hmm how's that possible since they both call the very same function to do the save? Or do you mean those fields do not trigger an autosave? Yes, I see, because they don't use the same form components.

I actually meant the DUO code adding/removing did not trigger autosave - I didn't test the answer fields, but they should too trigger the save.

File: src/cljs/rems/application.cljs [hunk 1]
```diff
@@ -224,11 +225,38 @@
...
+                      (apply (handle-validations! description application {:on-success #(do (rf/dispatch [::set-autosaving false])
+                                                                                            (flash-message/show-quiet-success! :actions-form-flash [text :t.form/autosave-confirmed]))
+                                                                           :default-success? false
+                                                                           :focus? false})
+                             [response])))
+       {:db (-> db
+                (assoc ::autosaving true)
+                (assoc-in [::edit-application :validation :errors] nil)
+                (assoc-in [::edit-application :validation :warnings] nil))})
+     {})))
```
Comment (aatkin) [inline src/cljs/rems/application.cljs:258]: one idea for shortening the function, if fetching application is not critical to perform prior to validations

```clojure
(->> (handle-validations! description
                          application
                          {:on-success (fn []
                                         (rf/dispatch [::set-autosaving false])
                                         (flash-message/show-quiet-success! :actions-form-flash
                                                                            [text :t.form/autosave-confirmed])
                                         ;; equal to (rf/dispatch [::fetch-application (:application/id application) false])
                                         (reload! (:application/id application) false))
                           :default-success? false
                           :focus? false})
     (save-draft! description application edit-application))
```

File: test/clj/rems/test_browser.clj [hunk 2]
```diff
@@ -507,13 +511,19 @@
...
 
-        (testing "save draft succesfully"
-          (btu/scroll-and-click :save)
-          (is (btu/eventually-visible? :status-success)))
+        (testing "saved manually or automatically"
+          (if (:enable-autosave env)
+            (is (btu/eventually-visible? {:id :status-success :fn/text "Application is saved."}))
+            (do
+              (btu/scroll-and-click :save)
+              (is (btu/eventually-visible? {:id :status-success})))))
```
Comment (Macroz) [inline test/clj/rems/test_browser.clj:519]: Let's do both. We can set the autosave on and off, and try that both work.

File: src/cljs/rems/application.cljs (hunk 1 above)
Comment (Macroz) [reply src/cljs/rems/application.cljs:258]: This is now included in #3037

**Review by aatkin — APPROVED** [review]
LGTM! :+1:

**Pull request merged**